DSEGOG-10 sticky column headers and ID column, plus date/time selector

sam-glendenning opened the pull request:
## Description
This PR does a lot of things :

- made column headers sticky, such that when the table is scrolled vertically, they remain in the viewport. I've simulated this by increasing the number of results per page and setting a maximum height on the table container
- made the ID column sticky, such that when the table is scrolled horizontally, the entire column and header remains in the viewport. You can simulate this by adding more columns than can fit in the viewport together
- added a date/time filter option above the table, separate from the component itself. These contain two date ranges for both start date/time and end date/time. Whenever one is either successfully filled with a valid date or cleared, the onChange method is fired, which updates the record query
- made the ID column visible at all times, with no option to remove it as per the issue details
- made the ID column non-draggable since it always needs to be on the left of the table
- recreated a lot of tests based on the above ID column changes
- added some more Cypress tests and generally tried to bump the global coverage in unit tests. _Check snapshots carefully_ as there have been many changes made and we want to make sure we get the expected results

I also noticed a bug on Friday afternoon that was present before starting this issue. When a column is added or removed, the current widths of all columns are put back to their default, erasing all user customised table sizing. I have not issued a fix for this in this PR as it's late on Friday and I wanted to get this reviewed by the time I get from leave. Either we fix it here, if it's easily doable, or we create an issue for it and attack it later.

## Testing instructions
I've tried my best to make sure everything plays together via unit tests but the amount of features in this table is quickly stacking up and will continue to lead to more unintended consequences (see above note about column resizing bug). So please test new functionality with as much existing functionality as possible.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking
Closes DSEGOG-10

Commit:
DSEGOG-10 removing Draggable element if column header is ID

This may not be the best solution in terms of ensuring we're dealing with the ID column but it'll do for now

Commit:
DSEGOG-10 providing height to column headers

Noticed a bug where a column header without a close icon (e.g. ID) had a smaller height than a column header that did. So I grabbed the correct height and applied it across the entire row to ensure consistency

Commit:
Merge branch 'DSEGOG-8-user-can-reorder-columns' into DSEGOG-10-sticky-columns

Commit:
DSEGOG-10 fix bug with ID header width

ID header can't be closed so we need to not show the close icon. However, this messed with the header width and made it out of sync of the rest of the column. A simple solution to combat this, instead of messing with column widths, was to hide the icon's visibility while still rendering it

Commit:
DSEGOG-10 ID column now appears by default

Opens when the table is first mounted and always appears on the left. I think this assumes an ID is always available? Which should be a reasonable assumption anyway? Not sure, this may need to change down the line

Commit:
DSEGOG-10 adding date-time input box

This adds the ability for the user to select a date and time to add to their query, with a date-time range for both start and end date.

Commit:
DSEGOG-10 adding date filter labels

Commit:
DSEGOG-10 making ID column sticky

After messing around with libraries and plugins for an eternity, in the end I found some simple CSS that seems to do the job. If we detect we're in the ID column, just plug it in with the other CSS

Commit:
DSEGOG-10 making column headers sticky

This just needed a little CSS and some clever z-index calculations. To simulate scrolling, I've added more results per page and limited the height of the table container

Commit:
Merge branch 'develop' into DSEGOG-10-sticky-columns

Commit:
DSEGOG-10 getting totally rid of any unnecessary padding

This should finally get rid of that pesky issue where columns were wider than they should have been

Commit:
DSEGOG-10 adding and updating tests

Commit:
DSEGOG-10 fixing date change call bug

The date change event was being called whenever the user entered the same date again. Fixed this by adding in a check to see if it matches what it previously was and doesn't call the event if so

Commit:
DSEGOG-10 fixing dateTimeInput snapshot and additional test

Commit:
DSEGOG-10 adding some table e2e tests

Commit:
DSEGOG-10 fixing clearing date filter not calling onChange

Clearing the date-time filter meant onChange was not called. This has now been fixed

louise-davies commented on src/table/table.component.tsx:
```diff
@@ -151,29 +180,30 @@ const Table = React.memo((props: TableProps): React.ReactElement => {
+                <MuiTableHead
+                  sx={{
+                    position: 'sticky',
+                    background: 'white', // theme colour later on
```
Can we add a TODO to this, as otherwise this comment could be forgotten about, whereas we know to search for TODO

louise-davies reviewed (commented):
Out of curiosity, can I ask why the date selector stuff was added to this PR? I think that stuff falls under the "Search" epic but if it's needed now then it's fine. Note that I think we only need two datepickers - as each shot has only 1 timestamp associated with it we just need a start and end picker for the range. This differs to DG where there were seperate columns in the DB which could have individual range filters applied to them

louise-davies commented on src/table/headerRenderers/dataHeader.component.tsx:
```diff
@@ -91,84 +91,99 @@ const DataHeader = (props: DataHeaderProps): React.ReactElement => {
+            setPermitDragging(true);
+          }}
+        >
+          {/* If this is an ID header, remove icon visibility but still render it
+          This ensures header widths remain consistent */}
```
The problem with this is that the ID column seemingly for no reason covers up the column name when resized very small. I've got a fix for this by using `display: none` and adjusting the `40px` value above

sam-glendenning replied:
That seems to work better, thanks for that

louise-davies commented on src/table/table.component.tsx:
```diff
@@ -186,27 +216,42 @@ const Table = React.memo((props: TableProps): React.ReactElement => {
                                   const { key, ...otherHeaderProps } =
                                     column.getHeaderProps();
 
+                                  const dataKey = column.render('id') as string;
```
I changed this in my other PR but I can ask it here - why do we use `column.render('id')` and not just `column.id`?

sam-glendenning replied:
I _think_ that was a remnant from a tutorial I followed whereby they were rendering 'id' as a basic text element. But it makes sense here to just use column.id so I'll change it

louise-davies committed:
Fix small resizing bug that caused cell dividers to disappear

louise-davies committed:
DSEGOG-10 let ID column resize nicely without close button spacing
- also refactor usage of TableCellContent to be more React typical

sam-glendenning commented:
> Out of curiosity, can I ask why the date selector stuff was added to this PR? I think that stuff falls under the "Search" epic but if it's needed now then it's fine. Note that I think we only need two datepickers - as each shot has only 1 timestamp associated with it we just need a start and end picker for the range. This differs to DG where there were seperate columns in the DB which could have individual range filters applied to them

So I did it like this because I thought allowing for a range selector for the date was more convenient for the end user and the API would allow for it in the same way with less than and greater than comparisons. I didn't have a look at the tasks in the Search epic so was just copying DG's method. If this isn't reflective of what the final product will be, I can go back to a basic start and end date selection.

And as for why I did it in the first place - it was included in the issue title and description but there's more to it overall that means it probably belongs in another epic/issue. Will adjust what I have so far to include only basic date-time selection and merge that.

Commit:
DSEGOG-10 switching to column.id, adding TODO to comments

Commit:
DSEGOG-10 removing unnecessary date-time inputs

As there is only one timestamp, we don't need a range for both a start and an end date so this is being removed.

Also improving the look of the date-time selector so that it matches the design in the Balsamiq wireframes

Commit:
DSEGOG-10 updating unit tests to account for new date-time filter

louise-davies commented:
@sam-glendenning We spoke about sticky columns in the quarterly report meeting and I got some opinions re: this requirement - see my comment here: https://stfc.atlassian.net/browse/DSEGOG-10?focusedCommentId=24900

Also, the merge conflicts need resolving 😅

Commit:
Merge branch 'develop' into DSEGOG-10-sticky-columns

Commit:
DSEGOG-10 fixing bug with sticky column not opening

This was due to the ID column not being "available" to display yet. This was because, although the data was received from the backend, it was still being transformed for display. As such, availableColumns was empty when the ID column was attempted to be opened

Commit:
Merge branch 'develop' into DSEGOG-10-sticky-columns

Commit:
DSEGOG-10 updating yarn dependencies

Commit:
Revert "DSEGOG-10 updating yarn dependencies"

This reverts commit b7a5388ee54b15a293f174b53cfb47dff0f44346.

Commit:
Revert "Merge branch 'develop' into DSEGOG-10-sticky-columns"

This reverts commit 2683284f73c5cf218dc30a0e29f05d27d376d2f1, reversing
changes made to 8c63fc7935577a77352e27c5a02229448b0bb903.

sam-glendenning commented:
I will likely abandon this PR in favour of [this new one](https://github.com/ral-facilities/operationsgateway/pull/13) instead. I managed to completely bork the yarn.lock file when trying to fix merge conflicts caused by adding Renovate to the repo. Since the issue has been updated with [new information](https://stfc.atlassian.net/browse/DSEGOG-10?focusedCommentId=24900) anyway, I thought I might as well make a fresh PR to address those changes anyway.

I'll leave this open for now, with a view to close it if/when the other is merged.

louise-davies commented:
Closed in favour of #13
